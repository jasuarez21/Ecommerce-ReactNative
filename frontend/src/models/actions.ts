import { Moto } from "./moto";
import { User } from "./user";

export type MotoActions = 
    | {type: 'GET_MOTOS'; data: Array<Moto>}
    | {type: 'ADD_TO_CART'; moto: Moto}
    | {type: 'LOAD_CART';}
    | {type: 'DELETE_CART'; moto: Moto}
    | {type: 'LOG_IN'; data: User}
    | {type: 'LOG_OUT'; data: User}
    | {type: 'SIGN_UP'; data: User}
    | {type: 'UPDATE_CART'}
    | {type: 'DELETE_ACCOUNT'}
    | {type: 'ERROR'}