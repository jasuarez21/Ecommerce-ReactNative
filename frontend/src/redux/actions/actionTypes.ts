const actionTypes = {
  GET_MOTOS: 'GET_MOTOS',
  LOG_IN: 'LOG_IN',
  LOG_OUT: 'LOG_OUT',
  SIGN_UP: 'SIGN_UP',
  ADD_TO_CART: 'ADD_TO_CART',
  LOAD_CART: 'LOAD_CART',
  DELETE_CART: 'DELETE_CART',
  UPDATE_CART: 'UPDATE_CART',
  SAVE_PRODUCT: 'SAVE_PRODUCT',
  DELETE_ACCOUNT: 'DELETE_ACCOUNT',
  ERROR: 'ERROR'
};

export default actionTypes;
