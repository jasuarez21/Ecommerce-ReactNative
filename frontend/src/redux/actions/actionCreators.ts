import { User } from './../../models/user';
import axios from 'axios';
import {URL_MOTOS, URL_LOGIN, URL_LOGOUT, URL_SIGNUP, URL_DELETE_ACCOUNT, URL_SAVE_PRODUCTS} from '@env';
import actionTypes from './actionTypes';
import {Moto} from '../../models/moto';
import {Dispatch} from 'redux';
import {MotoActions} from '../../models/actions';
import { Order } from '../../models/order';

export function GetMotos() {
  return async (dispatch: Dispatch) => {
      try {   
      const {data} = await axios.get(URL_MOTOS);
      dispatch({
        type: actionTypes.GET_MOTOS,
        data,
      });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}



export function login(email: string, password: string) {
  return async (dispatch: Dispatch) => {
      try {   
        const {data} = await axios.post(URL_LOGIN, {email, password});
        dispatch({
          type: actionTypes.LOG_IN,
          data,
        });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}

export function logout() {
  return async (dispatch: Dispatch) => {
      try {   
        const {data} = await axios.post(URL_LOGOUT);
        dispatch({
          type: actionTypes.LOG_OUT,
          data,
        });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}


export function signUp(email: string, password: string, alias: string, avatar: string) {
  return async (dispatch: Dispatch) => {
      try {   
        const {data} = await axios.post(URL_SIGNUP, {email, password, alias, avatar});
        dispatch({
          type: actionTypes.SIGN_UP,
          data,
        });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}

export function addToCart(moto: Moto): MotoActions {
  return {
    type: 'ADD_TO_CART',
    moto
  };
}

export function loadCart(): MotoActions {
  return {
    type: 'LOAD_CART'
  };
}

export function deleteFromCart(moto: Moto): MotoActions {
  return {
    type: 'DELETE_CART',
    moto,
  };
}

export function reloadCart(moto: Array<Moto>){
  return async (dispatch: Dispatch) => {
    try {   
      for(let i = 0; i < moto.length; i++){
        await axios.post(`${URL_MOTOS}/${moto[i]._id}`, moto[i]);
        dispatch({
          type: actionTypes.UPDATE_CART
        });
      }
  } catch (error) {
    dispatch({
      type: actionTypes.ERROR,
    });
  }
}
}

export function saveOrder(user: User, cart: Order, token: string){
  user.cart.push(cart);
  const config = { 
    headers: {
        authorization: `Bearer ${token}`,
    }
  }
    return async (dispatch: Dispatch) => {
      try {   
          await axios.post(URL_SAVE_PRODUCTS, {_id: user._id, cart: user.cart}, config);
          dispatch({
            type: actionTypes.SAVE_PRODUCT
          });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}

export function deleteAccount(user: User, token: string){
    return async (dispatch: Dispatch) => {
      try {   
          await axios.delete(URL_DELETE_ACCOUNT, {headers: {
              authorization: `Bearer ${token}`,
          },
              data:{_id: user._id}
          });
          dispatch({
            type: actionTypes.DELETE_ACCOUNT
          });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}