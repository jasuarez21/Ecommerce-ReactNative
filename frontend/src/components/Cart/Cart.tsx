import React from 'react';
import {Moto} from '../../models/moto';
import {RootState} from '../../redux/store/index'
import {StackScreenProps} from '@react-navigation/stack';
import {useSelector, useDispatch} from 'react-redux';
import {Text, SafeAreaView, View, ScrollView, Image, ImageBackground, TouchableHighlight, StyleSheet} from 'react-native';
import { deleteFromCart } from '../../redux/actions/actionCreators';

interface Props extends StackScreenProps<any, any> {}

const Cart = ({navigation}: Props) => {
  const cart = useSelector((store:RootState) => store.cart)
  const user = useSelector((store:RootState) => store.user);
  const dispatch = useDispatch();
  let precioTotal = 0;
  return (
    <ImageBackground
    style={styles.background}
    source={{
      uri: 'https://www.motoblouz.com/enjoytheride/wp-content/uploads/2018/10/cosmo_connected_nuit.jpg'
    }}>
       <SafeAreaView>
         <ScrollView>
     <Image
        style={styles.logo}
        source={{
          uri: 'https://i.postimg.cc/FF2Y5GJd/logo-small.png',
        }}
      />
    { cart.map((moto: Moto) =>(
      <View style={styles.componentMoto} key={moto._id}>
          <View style={styles.photoContainer}>
          <Image
                        style={styles.photo}
                        source={{
                          uri: `${moto.picture}`,
                        }}
                        />
          </View>
                        <View style={styles.dataContainer}>
                        <Text style={styles.cartInfo}> {moto?.marca} {moto?.modelo}</Text>
                              <Text style={styles.cartInfo}>Cantidad: {moto?.quantity} </Text>
                              <Text style={styles.precioMoto}>{moto.quantity * moto?.precio}€</Text>
                              <Text style={styles.increment}>{precioTotal += moto.quantity * moto?.precio}€</Text>
                        </View>
          <View style={styles.deleteContainer}>
          <TouchableHighlight
            onPress= {()=> {           
              dispatch(deleteFromCart(moto))
            }}
            >
            <Image
        testID= 'buttonDelete'
        style={styles.delete}
        source={{
          uri: 'https://i.postimg.cc/nhvmXCj2/icons8-xbox-x-50.png',
        }}
      />
      </TouchableHighlight>
      </View>
                <Text style={styles.price}>{precioTotal}</Text>
      </View>
    ))}
    </ScrollView>
    </SafeAreaView>
    <View style={styles.containerTotalPrice}>
      <Text style={styles.totalPrice}>TOTAL: {precioTotal} €</Text>
      { user.user !== undefined ? (
                <TouchableHighlight
                  onPress={()=> {
                    navigation.navigate('ContactForm')
                  }}
                >
                <Image
                        testID = 'buyButton'
                        style={styles.buy}
                        source={{
                          uri: 'https://i.postimg.cc/xCtp4npb/icons8-us-dollar-50.png',
                        }}
                      />  
                </TouchableHighlight>
      ): (
        <Text style={styles.login}>Inicia sesión para comprar</Text>
      )
}         
    </View>
    </ImageBackground>
  )
};
const styles = StyleSheet.create({
  deleteContainer: {
    marginRight: 10
  },
  precioMoto: {
    color: '#FFB800',
    fontSize: 24,
    marginTop: 10,
    marginLeft: 20
  },
  cartInfo: {
    fontSize: 18,
    color: 'black'
  },
  increment: {
    opacity: 0
  },
  photo: {
    width: 180,
    marginLeft: 15,
    height: 130,
    borderRadius: 25
  },
  photoContainer: {
    height: 150,
    width: 200,
    marginLeft: 23,
    marginTop: 10
  },
  dataContainer: {
    width: 150,
    height: 130,
    borderRadius: 15,
    backgroundColor: 'white',
    marginLeft: 20,
    padding: 20,
    marginTop: 10
  },
  componentMoto: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'center',
    marginTop: 10,
    width: 390,
    marginLeft: 15,
    borderTopColor: '#FFB800',
    borderTopWidth: 2,
  },
  background: {
    width: 600,
    height: 700
  },
  buy: {
    width: 30,
    height: 30
  },
  description: {
    display: 'flex',
    flexDirection: 'row'
  },
  login: {
    color: '#FFB800'
  },
  delete: {
    width: 30,
    height: 30,
    marginTop: 10
  },
  containerTotalPrice: {
    backgroundColor: '#464646',
    width: 415,
    height: 50,
    marginTop: 552,
    position: 'absolute',
    padding: 10,
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  totalPrice: {
    color: '#FFB800',
    fontSize: 20,
    alignSelf: 'center'
  },
  price:{
    opacity: 0
  },
  logo: {
    width: 300,
    height: 42,
    marginTop: 15,
    marginLeft: 10
  },
  scrollCart: {
    backgroundColor: 'white',
    width: 200,
    height: 170,
    borderRadius: 15,
    alignSelf: 'center',
    marginTop: 15
  }
});

export default (Cart);