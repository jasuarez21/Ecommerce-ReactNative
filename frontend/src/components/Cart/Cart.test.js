import { render, fireEvent } from '@testing-library/react-native';
import React from 'react';
import ConfigureStore from 'redux-mock-store';
import * as actions from '../../redux/actions/actionCreators';
import { Provider } from 'react-redux';
import Cart from './Cart';

describe('When invoked a Cart component', () => {
    const mockStore = ConfigureStore();
    const store = mockStore({
        cart: [{
          moto: {
              _id: '2',
              marca: 'Aprilia',
              modelo: 'SR',
              cc: 50, 
              precio: 550,
              descripcion: 'Moto comoda para la carretera',
              picture: 'img.com'
          }
        }],
        user: [{
            _id: '2',
            user : {
                alias: 'Anna',
                email: 'annita',
                password: 'annaaaza',
                avatar: 'img.com'
            }
        }
      ]
    });
    beforeEach(() => {
        jest.spyOn(actions, 'deleteFromCart').mockReturnValueOnce({ type: '' });
        jest.spyOn(actions, 'reloadCart').mockReturnValueOnce({ type: '' });
      });

    test('Should render a moto with detail', () => {
        const cart = render(<Provider store={store}><Cart /></Provider>)
        expect(cart).toMatchSnapshot();
    })
    test('Then login function is invoked', () => {
      const { getByTestId } = render(<Provider store={store}><Cart /></Provider>);
      const image = getByTestId('buttonDelete');
      fireEvent.press(image);
      expect(actions.deleteFromCart).toHaveBeenCalled();
    });
})

describe('When invoked a Cart component', () => {
  const mockStore = ConfigureStore();
  const store = mockStore({
      cart: [{
        moto: {
            _id: '2',
            marca: 'Aprilia',
            modelo: 'SR',
            cc: 50, 
            precio: 550,
            descripcion: 'Moto comoda para la carretera',
            picture: 'img.com'
        }
      }],
      user: {
          _id: '2',
          user : {
              alias: 'Anna',
              email: 'annita',
              password: 'annaaaza',
              avatar: 'img.com'
          }
      }
  });
  let navigation;
  beforeEach(() => {
      jest.spyOn(actions, 'deleteFromCart').mockReturnValueOnce({ type: '' });
      jest.spyOn(actions, 'reloadCart').mockReturnValueOnce({ type: '' });
      navigation = {
        navigate: jest.fn()
      }
    });

  test('Should render a moto with detail', () => {
      const cart = render(<Provider store={store}><Cart /></Provider>)
      expect(cart).toMatchSnapshot();
  })
  test('Then login function is invoked', () => {
    const { getByTestId } = render(<Provider store={store}><Cart navigation={navigation}/></Provider>);
    const image = getByTestId('buyButton');
    fireEvent.press(image);
    expect(navigation.navigate).toHaveBeenCalled();
  });
})