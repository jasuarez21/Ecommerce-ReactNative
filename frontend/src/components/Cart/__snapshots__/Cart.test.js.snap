// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When invoked a Cart component Should render a moto with detail 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "height": 700,
      "width": 600,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://www.motoblouz.com/enjoytheride/wp-content/uploads/2018/10/cosmo_connected_nuit.jpg",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": 700,
          "width": 600,
        },
        undefined,
      ]
    }
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
  >
    <RCTScrollView>
      <View>
        <Image
          source={
            Object {
              "uri": "https://i.postimg.cc/FF2Y5GJd/logo-small.png",
            }
          }
          style={
            Object {
              "height": 42,
              "marginLeft": 10,
              "marginTop": 15,
              "width": 300,
            }
          }
        />
        <View
          style={
            Object {
              "borderTopColor": "#FFB800",
              "borderTopWidth": 2,
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
              "marginLeft": 15,
              "marginTop": 10,
              "width": 390,
            }
          }
        >
          <View
            style={
              Object {
                "height": 150,
                "marginLeft": 23,
                "marginTop": 10,
                "width": 200,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "undefined",
                }
              }
              style={
                Object {
                  "borderRadius": 25,
                  "height": 130,
                  "marginLeft": 15,
                  "width": 180,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "backgroundColor": "white",
                "borderRadius": 15,
                "height": 130,
                "marginLeft": 20,
                "marginTop": 10,
                "padding": 20,
                "width": 150,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                }
              }
            >
               
               
            </Text>
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                }
              }
            >
              Cantidad: 
               
            </Text>
            <Text
              style={
                Object {
                  "color": "#FFB800",
                  "fontSize": 24,
                  "marginLeft": 20,
                  "marginTop": 10,
                }
              }
            >
              NaN
              €
            </Text>
            <Text
              style={
                Object {
                  "opacity": 0,
                }
              }
            >
              NaN
              €
            </Text>
          </View>
          <View
            style={
              Object {
                "marginRight": 10,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <Image
                source={
                  Object {
                    "uri": "https://i.postimg.cc/nhvmXCj2/icons8-xbox-x-50.png",
                  }
                }
                style={
                  Object {
                    "height": 30,
                    "marginTop": 10,
                    "width": 30,
                  }
                }
                testID="buttonDelete"
              />
            </View>
          </View>
          <Text
            style={
              Object {
                "opacity": 0,
              }
            }
          >
            NaN
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
  <View
    style={
      Object {
        "backgroundColor": "#464646",
        "display": "flex",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "marginTop": 552,
        "padding": 10,
        "position": "absolute",
        "width": 415,
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#FFB800",
          "fontSize": 20,
        }
      }
    >
      TOTAL: 
      NaN
       €
    </Text>
    <Text
      style={
        Object {
          "color": "#FFB800",
        }
      }
    >
      Inicia sesión para comprar
    </Text>
  </View>
</View>
`;

exports[`When invoked a Cart component Should render a moto with detail 2`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "height": 700,
      "width": 600,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://www.motoblouz.com/enjoytheride/wp-content/uploads/2018/10/cosmo_connected_nuit.jpg",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": 700,
          "width": 600,
        },
        undefined,
      ]
    }
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
  >
    <RCTScrollView>
      <View>
        <Image
          source={
            Object {
              "uri": "https://i.postimg.cc/FF2Y5GJd/logo-small.png",
            }
          }
          style={
            Object {
              "height": 42,
              "marginLeft": 10,
              "marginTop": 15,
              "width": 300,
            }
          }
        />
        <View
          style={
            Object {
              "borderTopColor": "#FFB800",
              "borderTopWidth": 2,
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
              "marginLeft": 15,
              "marginTop": 10,
              "width": 390,
            }
          }
        >
          <View
            style={
              Object {
                "height": 150,
                "marginLeft": 23,
                "marginTop": 10,
                "width": 200,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "undefined",
                }
              }
              style={
                Object {
                  "borderRadius": 25,
                  "height": 130,
                  "marginLeft": 15,
                  "width": 180,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "backgroundColor": "white",
                "borderRadius": 15,
                "height": 130,
                "marginLeft": 20,
                "marginTop": 10,
                "padding": 20,
                "width": 150,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                }
              }
            >
               
               
            </Text>
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                }
              }
            >
              Cantidad: 
               
            </Text>
            <Text
              style={
                Object {
                  "color": "#FFB800",
                  "fontSize": 24,
                  "marginLeft": 20,
                  "marginTop": 10,
                }
              }
            >
              NaN
              €
            </Text>
            <Text
              style={
                Object {
                  "opacity": 0,
                }
              }
            >
              NaN
              €
            </Text>
          </View>
          <View
            style={
              Object {
                "marginRight": 10,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <Image
                source={
                  Object {
                    "uri": "https://i.postimg.cc/nhvmXCj2/icons8-xbox-x-50.png",
                  }
                }
                style={
                  Object {
                    "height": 30,
                    "marginTop": 10,
                    "width": 30,
                  }
                }
                testID="buttonDelete"
              />
            </View>
          </View>
          <Text
            style={
              Object {
                "opacity": 0,
              }
            }
          >
            NaN
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
  <View
    style={
      Object {
        "backgroundColor": "#464646",
        "display": "flex",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "marginTop": 552,
        "padding": 10,
        "position": "absolute",
        "width": 415,
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#FFB800",
          "fontSize": 20,
        }
      }
    >
      TOTAL: 
      NaN
       €
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Image
        source={
          Object {
            "uri": "https://i.postimg.cc/xCtp4npb/icons8-us-dollar-50.png",
          }
        }
        style={
          Object {
            "height": 30,
            "width": 30,
          }
        }
        testID="buyButton"
      />
    </View>
  </View>
</View>
`;
